kotlinpoet/src/main/java/com/squareup/kotlinpoet/Util.kt

```diff
@@ -207,18 +211,22 @@ private fun String.failIfEscapeInvalid() {
 }
 
 internal fun String.escapeIfNecessary(validate: Boolean = true): String {
-  val escapedString = escapeIfNotJavaIdentifier().escapeIfKeyword()
+  val escapedString = escapeIfNotJavaIdentifier().escapeIfKeyword().escapeIfAllowedCharacter()
   if (validate) {
     escapedString.failIfEscapeInvalid()
   }
   return escapedString
 }
 
-private fun String.escapeIfKeyword() = if (isKeyword) "`$this`" else this
+private fun String.alreadyEscaped() = startsWith("`") && endsWith("`")
```

Comment:
Added this since we're chaining escapes like `escapeIfNotJavaIdentifier().escapeIfKeyword().escapeIfAllowedCharacter()`, there are chances of double escaping a string if two of the conditions meet. 

Happy to know if there's a more idiomatic way of representing this (i.e execute next statement if this predicate matches)

Reply:
I think it's fine, we could also just refactor the "escape" functions to return booleans, `||` them together and escape based on the value.